### PR TITLE
RT600 PWM support

### PR DESCRIPTION
**Pull request by mmahadevan108**
Tested with `tests/drivers/pwm/pwm_api`. Used a scope to confirm the PWM signal.

### CODE REVIEW

**Review by galak — DISMISSED** [review]
Add to `tests/drivers/build_all/pwm`

**Comment by mmahadevan108** [comment]
> Add to `tests/drivers/build_all/pwm`

I have updated.

**Comment by mmahadevan108** [comment]
Anything else needed for this PR?

File: dts/bindings/pwm/nxp,lpc-pwm.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+compatible: "nxp,lpc-pwm"
+
+include: [pwm-controller.yaml, base.yaml]
+
+properties:
+    prescaler:
+      type: int
+      required: false
+      default: 1
```
Comment (mbolivar-nordic) [inline dts/bindings/pwm/nxp,lpc-pwm.yaml:14]: All uses of default must be justified in the description as documented in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#default-values-for-properties.

File: tests/drivers/pwm/pwm_api/src/test_pwm.c [hunk 2]
```diff
@@ -185,4 +187,5 @@ void test_pwm_cycle(void)
 	/* Period : Pulse (64000 : 0), unit (cycle). Voltage : 0V */
 	zassert_true(test_task(DEFAULT_PWM_PORT, DEFAULT_PERIOD_CYCLE,
 				0, UNIT_CYCLES) == TC_PASS, NULL);
+
```
Comment (mbolivar-nordic) [inline tests/drivers/pwm/pwm_api/src/test_pwm.c:190]: Whitespace nit

File: west.yml [hunk 3]
```diff
@@ -100,7 +100,7 @@ manifest:
       revision: f49bd1354616fae4093bf36e5eaee43c51a55127
       path: tools/net-tools
     - name: hal_nxp
-      revision: 0d11138724959e1162777d9206f841ccdf64348e
+      revision: pull/90/head
```
Comment (mbolivar-nordic) [inline west.yml:103]: Any progress on getting this merged?

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
> Anything else needed for this PR?

- Needs a rebase
- The HAL update needs to be merged and west.yml updated with the final SHA
- The devicetree binding uses `default` without justification

File: tests/drivers/pwm/pwm_api/src/test_pwm.c (hunk 2 above)
Comment (mmahadevan108) [reply tests/drivers/pwm/pwm_api/src/test_pwm.c:190]: Thank you for spotting this. I removed this change.

File: dts/bindings/pwm/nxp,lpc-pwm.yaml (hunk 1 above)
Comment (mmahadevan108) [reply dts/bindings/pwm/nxp,lpc-pwm.yaml:14]: Thank you, I have updated the property description

File: west.yml (hunk 3 above)
Comment (mmahadevan108) [reply west.yml]: I have pinged the team for their review comments, will update this once the PR is merged.

Comment (MaureenHelm) [reply west.yml]: Done

Comment (mmahadevan108) [reply west.yml]: This has been updated.

File: boards/arm/mimxrt685_evk/doc/index.rst [hunk 4]
```diff
@@ -129,12 +131,12 @@ functionality of a pin.
...
 | PIO0_21 | I2S             | I2S TX SCK                 |
 +---------+-----------------+----------------------------+
+| PIO0_9  | I2S             | I2S DATAIN                 |
++---------+-----------------+----------------------------+
 | PIO0_29 | USART           | USART TX                   |
 +---------+-----------------+----------------------------+
 | PIO0_30 | USART           | USART RX                   |
 +---------+-----------------+----------------------------+
-| PIO0_9  | I2S             | I2S DATAIN                 |
-+---------+-----------------+----------------------------+
```
Comment (MaureenHelm) [inline boards/arm/mimxrt685_evk/doc/index.rst:137]: This change is not related to PWM

File: boards/arm/mimxrt685_evk/pinmux.c [hunk 5]
```diff
@@ -444,7 +444,7 @@ static int mimxrt685_evk_pinmux_init(const struct device *dev)
 #endif
 
 #if DT_NODE_HAS_STATUS(DT_NODELABEL(flexspi), okay) && CONFIG_FLASH
-const uint32_t port1_pin11_config = (/* Pin is configured as FLEXSPI0B_DATA0 */
+	uint32_t port1_pin11_config = (/* Pin is configured as FLEXSPI0B_DATA0 */
```
Comment (MaureenHelm) [inline boards/arm/mimxrt685_evk/pinmux.c:447]: Please move all the whitespace fixes into a separate commit

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: boards/arm/mimxrt685_evk/doc/index.rst (hunk 4 above)
Comment (mmahadevan108) [reply boards/arm/mimxrt685_evk/doc/index.rst:137]: For some reason the I2S related items were not grouped together. I made this change while adding the PWM information.

File: boards/arm/mimxrt685_evk/pinmux.c (hunk 5 above)
Comment (mmahadevan108) [reply boards/arm/mimxrt685_evk/pinmux.c]: I will make a separate PR

Comment (mmahadevan108) [reply boards/arm/mimxrt685_evk/pinmux.c]: PR https://github.com/zephyrproject-rtos/zephyr/pull/36785 created to address the other issues with this file.

File: boards/arm/mimxrt685_evk/doc/index.rst (hunk 4 above)
Comment (MaureenHelm) [reply boards/arm/mimxrt685_evk/doc/index.rst:137]: It's ok to move the I2S items together, just do it in a separate commit

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Can we rename this driver from `mcux_lpc` to `mcux_sctimer`? This better reflects the peripheral hardware IP and leaves room for the possibility of another PWM driver in the LPC family. I've tried to limit use of `mcux_lpc` to only cases where there are hardware IPs with the same name in different SoC families.

File: dts/bindings/pwm/nxp,lpc-pwm.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# Copyright (c) 2017, NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: NXP LPC PWM
+
+compatible: "nxp,lpc-pwm"
```
Comment (galak) [inline dts/bindings/pwm/nxp,lpc-pwm.yaml:6]: why "nxp,lpc-pwm".  seems like we should have sctimer in the name.

File: boards/arm/mimxrt685_evk/doc/index.rst (hunk 4 above)
Comment (mmahadevan108) [reply boards/arm/mimxrt685_evk/doc/index.rst:137]: Done, PR https://github.com/zephyrproject-rtos/zephyr/pull/36809 created

File: dts/bindings/pwm/nxp,lpc-pwm.yaml (hunk 6 above)
Comment (mmahadevan108) [reply dts/bindings/pwm/nxp,lpc-pwm.yaml]: I have renamed the driver to `pwm_mcux_sctimer.c` and changed this to `nxp_sctimer_pwm`

**Comment by mmahadevan108** [comment]
> Can we rename this driver from `mcux_lpc` to `mcux_sctimer`? This better reflects the peripheral hardware IP and leaves room for the possibility of another PWM driver in the LPC family. I've tried to limit use of `mcux_lpc` to only cases where there are hardware IPs with the same name in different SoC families.

I have renamed the driver to `pwm_mcux_sctimer.c` and have renamed the configs, device tree related items and the function/variable names in the driver.

File: boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts [hunk 7]
```diff
@@ -19,6 +19,8 @@
 		led1 = &blue_led;
 		led2 = &red_led;
 		usart-0 = &flexcomm0;
+		/* For pwm test suites */
+		pwm-0 = &sc_timer;
```
Comment (MaureenHelm) [inline boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:23]: Can the LEDs on this board be driven by a PWM? If so, please add pwm-led aliases to enable `samples/basic/rgb_led` and `samples/drivers/led_pwm`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
my comments are addressed

**Comment by LeBlue** [comment]
There is a major issue with the current implementation of this driver. The peripheral has a limitation that is also documented in the hal API that is used to set/change the PWM period:
[SCTIMER_SetupPwm](https://github.com/zephyrproject-rtos/hal_nxp/blob/6d34c83c485865c25f90456d00e054947ca54978/mcux/drivers/imxrt6xx/fsl_sctimer.c#L240)

In short: the PWM period is not (really) reconfigurable (hal API/implementation limitation) and shared between all channels (peripheral limitation), only the pulse width is changeble per channel. Changing it for one channel changes the period for all. The driver does not prevent that, but it should.

I have tested it with the lpcxresso55s28 board (should be the same IP core) and the blinky_pwm example that fails because the limitation (driver errors out after changing period for more than numer of available channels.)

To make the driver usable for now with the limitation and minimal changes, the period could be fixed after first configuration.

**Comment by MaureenHelm** [comment]
- Rebased and resolved merge conflicts
- Added a new commit to enable PWM LEDs on the board. I tested with `samples/basic/fade_led` and found an issue when the pulse width is set zero. I'm not sure what the right resolution is yet, but the sample works correctly if I comment out the line that restarts the timer in the 0% duty cycle case.

> In short: the PWM period is not (really) reconfigurable (hal API/implementation limitation) and shared between all channels (peripheral limitation), only the pulse width is changeble per channel. Changing it for one channel changes the period for all. The driver does not prevent that, but it should.

It's not unusual for all channels to share the same period, some drivers log a warning: https://github.com/NXPmicro/zephyr/blob/b8820da4e6341ff4fe89fc7e1a23c1995ece1b43/drivers/pwm/pwm_mcux_ftm.c#L111-L114

> I have tested it with the lpcxresso55s28 board (should be the same IP core) and the blinky_pwm example that fails because the limitation (driver errors out after changing period for more than numer of available channels.)

Thanks for testing. I see the same behavior on the `mimxrt685_evk_cm33` board.

**Comment by mmahadevan108** [comment]
Thank you. I have added a fix to the driver for the case when the dutycyle is zero to check if the PWM has been setup before starting the timer.

**Review by MaureenHelm — APPROVED** [review]
